pkg/controller/datasource-controller.go

```diff
@@ -126,19 +128,72 @@ func NewDataSourceController(mgr manager.Manager, log logr.Logger, installerLabe
 	if err != nil {
 		return nil, err
 	}
-	if err := addDataSourceControllerWatches(mgr, DataSourceController, log); err != nil {
+	if err := addDataSourceControllerWatches(mgr, DataSourceController); err != nil {
 		return nil, err
 	}
 	log.Info("Initialized DataSource controller")
 	return DataSourceController, nil
 }
 
-func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller) error {
 	if err := cdiv1.AddToScheme(mgr.GetScheme()); err != nil {
 		return err
 	}
-	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: func(e event.UpdateEvent) bool { return !sameDataSourcePvc(e.ObjectOld, e.ObjectNew) },
+		},
+	); err != nil {
+		return err
+	}
+
+	const dataSourcePvcField = "spec.source.pvc"
+
+	getKey := func(namespace, name string) string {
+		return namespace + "." + name
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &cdiv1.DataSource{}, dataSourcePvcField, func(obj client.Object) []string {
+		if pvc := obj.(*cdiv1.DataSource).Spec.Source.PVC; pvc != nil {
+			return []string{getKey(pvc.Namespace, pvc.Name)}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	mapToDataSource := func(obj client.Object) (reqs []reconcile.Request) {
+		var dataSources cdiv1.DataSourceList
+		matchingFields := client.MatchingFields{dataSourcePvcField: getKey(obj.GetNamespace(), obj.GetName())}
+		mgr.GetClient().List(context.TODO(), &dataSources, matchingFields)
```

Comment:
should we log any error here and return?

Reply:
added